plutus-ir/src/Language/PlutusIR/Transform/ThunkRecursions.hs

```diff
@@ -0,0 +1,174 @@
+{-# LANGUAGE FlexibleContexts  #-}
+{-# LANGUAGE LambdaCase        #-}
+{-# LANGUAGE OverloadedStrings #-}
+-- | Implements a PIR-to-PIR transformation that makes all recursive term definitions
+-- compilable to PLC. See Note [Thunking recursions] for details.
+module Language.PlutusIR.Transform.ThunkRecursions (thunkRecursionsTerm) where
+
+import           PlutusPrelude
+
+import           Language.PlutusIR
+import           Language.PlutusIR.MkPir
+import           Language.PlutusIR.Transform.Substitute
+
+import           Language.PlutusCore.Quote
+import qualified Language.PlutusCore.StdLib.Data.Unit   as Unit
+
+import           Control.Lens
+
+import           Data.List
+import qualified Data.Map                               as Map
+import           Data.Maybe
+import           Data.Traversable
+
+{- Note [Thunking recursions]
+Our fixpoint combinators in Plutus Core know how to handle mutually recursive values
+of *function type*. That is, we can define `map : List Int -> List Int` but *not*
+`map : forall a . List a -> List a`, because the latter has a universally
+qualified type, instead of a function type (although it is a function underneath).
+
+We could solve this problem for universally quantified values by lifting the forall
+out of the recursion. This is a bit like performing the following transformation:
+
+    map : forall a . List a -> List a
+    map f xs = case xs of
+        [] -> []
+        x:xs -> f x : map f xs
```

Comment:
It should be `map : forall a . (a -> a) -> List a -> List a` or `map : forall a b . (a -> b) -> List a -> List b`, I guess?

Reply:
:sweat_smile: you are a much better typechecker than my brain.